crates/state-read-vm/src/state_read.rs

```diff
@@ -141,12 +142,27 @@ where
     Ok(state_read.key_range(contract_addr, key, num_keys))
 }
 
-/// Write the given values to mutable state slots.
-fn write_values_to_state_slots<E>(
-    vm: &mut Vm,
-    slot_index: usize,
+/// Write the given values to memory.
+fn write_values_to_memory<E>(
+    mem_addr: usize,
     values: Vec<Vec<Word>>,
+    vm: &mut Vm,
 ) -> OpAsyncResult<(), E> {
-    vm.state_memory.store_slots_range(slot_index, values)?;
+    let values_len = Word::try_from(values.len()).map_err(|_| MemoryError::Overflow)?;
+    let index_len_pairs_len = values_len.checked_mul(2).ok_or(MemoryError::Overflow)?;
+    let mut mem_addr = Word::try_from(mem_addr).map_err(|_| MemoryError::IndexOutOfBounds)?;
+    let mut value_addr = mem_addr
+        .checked_add(index_len_pairs_len)
+        .ok_or(MemoryError::Overflow)?;
+    for value in values {
+        let value_len = Word::try_from(value.len()).map_err(|_| MemoryError::Overflow)?;
+        // Write the [index, len] pair.
+        vm.memory.store_range(mem_addr, &[value_addr, value_len])?;
+        // Write the value.
+        vm.memory.store_range(value_addr, &value)?;
+        // No need to check addition here as `store_range` would have failed.
+        value_addr += value_len;
+        mem_addr += 2;
+    }
```

Comment:
Nice. Maybe we should document this layout somewhere?

Reply:
This is documented in the updated asm.yml docs for the KeyRange ops, but I've also added a small example just to make it a bit clearer.